make schema a property on the model independent of prototype

mliszewski opened the pull request:
This PR is to address https://github.com/sequelize/sequelize/issues/4119.

I added some unit and integration tests to cover the behavior more thoroughly.

janmeier commented on test/integration/associations/belongs-to-many.test.js:
```diff
@@ -206,7 +206,11 @@ describe(Support.getTestDialectTeaser('BelongsToMany'), function() {
         return self.sequelize.createSchema('acme');
       }).then(function() {
-        return self.sequelize.sync({force: true});
+        return AcmeUser.sync({force: true});
```
Why the individual sync here? Sequelize should be able to figure out the right order

mliszewski replied:
The models that are attached to the sequelize object to do not have the schema set anymore since .schema no longer sets the option for all models but now creates a copy of the model with the $schema attribute set. sequelize.define will only addd the base model with the default schema. An alternative would be to set the schema attribute on the define but then that wouldn't be using .schema at all.

janmeier replied:
True :)

janmeier replied:
I'd prefer an `Promise.all([sync, sync, sync` though

janmeier commented on test/integration/model/schema.test.js:
```diff
@@ -0,0 +1,458 @@
+        return current.createSchema('schema_one').then(function() {
+          return current.createSchema('schema_two');
+        }).then(function() {
+          return self.RestaurantOne.sync({force: true});
```
return self.sequelize.sync()

And you don't need to `catch` - the test will fail if the chain is rejected

mliszewski replied:
Cannot use self.sequelize.sync b/c of my comment above.

janmeier commented on test/integration/model/schema.test.js:
```diff
@@ -0,0 +1,458 @@
+
+          return self.RestaurantTwo.create({
+            foo: 'test'
+          }).catch(function(err) {
```
No need to `catch`

janmeier commented on test/integration/model/schema.test.js:
```diff
@@ -0,0 +1,458 @@
+            });
+        });
+
+        it('should fail to insert data into schema_two using create', function() {
```
Why should it fail?

mliszewski replied:
It shouldn't. 

- [ ] Delete test

janmeier commented on test/integration/model/schema.test.js:
```diff
@@ -0,0 +1,458 @@
+        it('should fail to find schema_two object in schema_one', function() {
+          var self = this;
+
+          return self.RestaurantOne.findOne({where: {foo: 'two'}}).then(function(RestaurantObj) {
```
The tables are empty at this point, so what are we testing exactly :D ?

mliszewski replied:
These two tests should be moved to the 2 tests above them to verify I cannot get the data in the opposite schema.

- [ ] Combine opposite schema verification into create data tests

janmeier commented on test/integration/model/schema.test.js:
```diff
@@ -0,0 +1,458 @@
+              restaurauntModel = self.RestaurantOne.build({bar: 'one.2'});
+              return restaurauntModel.save();
+            }).then(function() {
+              restaurauntModel = self.RestaurantTwo.build({bar: 'two.1'});
```
`return self.RestaurantTwo.create({bar: 'two.1'});`

mliszewski replied:
This is purposefully done in two steps because before this PR a race condition existed wherein if you called save on a built model after calling .schema with a different schema it would save into the new schema.

mickhansen replied:
Wouldn't that case happen if the two saves happened in parallel? Which they don't here

mliszewski replied:
They don't have to be in parallel but out of order. In other words build.schema1, build.schema2, save.schema1 would fail. I do have a test for this explicitly later though so I'll clean these up.

janmeier commented on test/integration/model/schema.test.js:
```diff
@@ -0,0 +1,458 @@
+              });
+              return self.RestaurantOne.findAndCountAll();
+            }).then(function(restaurantsOne) {
+              expect(restaurantsOne).to.not.be.null;
```
This block is saying the same thing as the previous, but with `findAndCount` right?

mliszewski replied:
Yes. Just making sure all operations behave correctly.

janmeier replied:
Yeah, I realized so too after writing the comment ;)

janmeier commented on test/integration/model/schema.test.js:
```diff
@@ -0,0 +1,458 @@
+          var restaurauntModel = self.RestaurantOne.build({bar: 'one.1'});
+
+          return restaurauntModel.save().then(function() {
+            restaurauntModel = self.RestaurantTwo.build({bar: 'two.1'});
```
You can merge the creation into one with
```js
return Promise.all([
  self.RestaurantTwo.bulkCreate([
    { bar ... },
   { bar .... }
  ]),
  self.RestaurantOne.bulkCreate([
    { bar ... },
   { bar .... }
  ]),
]);
```

mliszewski replied:
I'm purposefully doing this in multiple steps to verify that built models maintain there proper $schema setting across calls as this was not the case with the original implementation because it was changing the schema option on the shared prototype.

janmeier replied:
Yeah I see your point - Might want to write a comment so the structure isn't removed in an overzealous refactor then ;)

janmeier commented on test/integration/model/schema.test.js:
```diff
@@ -0,0 +1,458 @@
+        });
+
+
+        it('should be able to insert and retrieve associated data into the table in schema_two', function() {
```
This test doesn't add anything new in regards to schemas  - the result is the same regardless of which schema you are using

mliszewski replied:
True.

- [ ] Remove test

janmeier commented on lib/associations/belongs-to-many.js:
```diff
@@ -464,6 +464,10 @@ BelongsToMany.prototype.injectGetter = function(obj) {
     }
 
+    if(options.hasOwnProperty('schema')){
+      model = model.schema(options.schema,options.schemaDelimiter || '');
```
No need to add the default `|| ''` I think - An empty value should be handled further down

mickhansen replied:
Note on styling, we generally have a space between `if` and `(`.
We generally also have a space after a comma in a options set, so `options.schema, options.schemaDelimiter`:

This applies to the rest of the code aswell.

janmeier commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -21,8 +21,10 @@ var QueryGenerator = {
-      , schema = (!param.options && param.schema) || (param.options && param.options.schema ? param.options.schema : undefined)
-      , schemaDelimiter = (param.options && param.options.schemaDelimiter ? param.options.schemaDelimiter : undefined);
+      , schema = param.$schema ||
+        (param.options && param.options.schema ? param.options.schema : undefined)
```
Is the `param.options` still needed? Shouldn't it always be `params.$schema`?

janmeier commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -1536,7 +1538,7 @@ var QueryGenerator = {
 
       /* Attributes for the right side */
-      , right = association.target
+      , right = include.model
```
?

mliszewski replied:
This honors a models being passed on the include. association.target is the model at the time of the association being created. If you try to do 
```javascript
 return User.findAll({
    where: {
        first_name: 'matt'
    },
    include: [{
        model: Comments.schema('new_schema'),
        as: 'comments'
    }]
}); 
```
without this code it will join the included table with the schema set on base model. Without this .schema is pretty worthless for includes because you would have to add a separate association for every schema.

janmeier commented on lib/instance.js:
```diff
@@ -642,14 +643,14 @@ Instance.prototype.save = function(options) {
 
         query = 'update';
-        args = [self, self.Model.getTableName(options), values, where, options];
+        args = [self, self.Model.getTableName(this.$options), values, where, options];
```
The options object is never used by `getTableName`

mickhansen replied:
Actually it is, we let users override it to provide user specific table names based on options passed to a call.
So it should still be `options`, not `this.$options`

mickhansen replied:
Same goes for the rest of the codebase where call options has been changed for model/instance options (which doesn't make a lot of sense to me actually, why would instance options be used?)

janmeier replied:
Ah yes - Maybe `this` then - since the schema is not longer stored in options?

mickhansen replied:
Should be `options` as before, it was not about schema it was about providing custom tables by overloading getTableName

mliszewski replied:
options was unused in getTableName. the code looked like this 
```javascript
Model.prototype.getTableName = function(options) { // testhint options:none
  return this.QueryGenerator.addSchema(this);
}
```
The issue here is that this.Model is attached to the prototype of the model and not the model instance itself. See https://github.com/sequelize/sequelize/blob/master/lib/model.js#L718-L719. Because of this the $schema and $schemaDelimiter attributes could be different than the values on the model that find was called on. An alternative here would be to do ``` this.Model.schema(this.$options.$schema,this.$options.$schemaDelimiter) ``` for the case where ```this.Model.$schema !== this.$options.$schema || this.Model.$schemaDelimiter !== this.$options.$schemaDelimiter ```. Does that work? If so I will revert back ALL of the options changes including making getTableName do nothing with options.

mickhansen replied:
`options` is used by end-users, as i wrote in my previous comment, changing it would mean a major bump and we're not looking to do that quite yet :)

`> The issue here is that this.Model is attached to the prototype of the model and not the model instance itself

Not sure what you mean? It's attached to the instance prototype which means its attached to ALL instances. (Not sure how it's relevant either).

janmeier commented on lib/model.js:
```diff
@@ -271,8 +271,6 @@ function conformInclude(include, self) {
 
     include = { model: model, association: include, as: include.as };
   } else if (include instanceof Model) {
-    model = include;
```
?

mliszewski replied:
This was unused.

janmeier commented on lib/model.js:
```diff
@@ -1051,7 +1054,16 @@ Model.prototype.schema = function(schema, options) { // testhint options:none
+      name: this.name,
+      $schema: options.$schema || this.$schema,
+      $schemaDelimiter: options.$schemaDelimiter || this.$schemaDelimiter
+    });
```
```js
return this.QueryGenerator.addSchema(_.assign({
      tableName: this.tableName,
      name: this.name,
      $schema: this.$schema,
      $schemaDelimiter: this.$schemaDelimiter
    }, options);
```
should cover both cases

janmeier commented on test/unit/model/schema.test.js:
```diff
@@ -0,0 +1,42 @@
+
+describe(Support.getTestDialectTeaser('Model'), function() {
+  if (current.dialect.supports.schemas) {
+    var Project = current.define('project'),
```
+1 for unit tests - nice! 

Might want to add one for delimiter as well

janmeier commented on lib/associations/belongs-to.js:
```diff
@@ -159,6 +159,10 @@ BelongsTo.prototype.injectGetter = function(instancePrototype) {
       }
     }
 
+    if(options.hasOwnProperty('schema')){
```
Remember to add API docs

janmeier commented:
Overall solid and well thought through work both in the test and code department @mliszewski !

mickhansen commented on lib/model.js:
```diff
@@ -978,9 +979,9 @@ Model.prototype.sync = function(options) {
+    return self.QueryInterface.createTable(self.getTableName(), attributes, options, self);
   }).then(function () {
-    return self.QueryInterface.showIndex(self.getTableName(options), options);
+    return self.QueryInterface.showIndex(self.getTableName(), options);
```
Restore `options`, see other comment.

mickhansen commented on lib/model.js:
```diff
@@ -1027,19 +1028,21 @@ Model.prototype.dropSchema = function(schema) {
  */
 Model.prototype.schema = function(schema, options) { // testhint options:none
-  this.options.schema = schema;
+  var self = Object.create(this);
```
Perhaps use `clone` rather than `self` since `self` usually is a reference to the original `this`.

mliszewski commented:
@mickhansen @janmeier I cleaned up some items based on your feedback. Below is a summary of my changes.

- Added API docs
- Added unit test for setting schema delimiter
- Removed/consolidated redundant test cases in schema test
- Reverted changes regarding ```options```

Let me know what you think when you get chance. Thanks!

mickhansen commented on lib/associations/belongs-to-many.js:
```diff
@@ -464,6 +465,10 @@ BelongsToMany.prototype.injectGetter = function(obj) {
 
+    if (options.hasOwnProperty('schema')) {
+      model = model.schema(options.schema, options.schemaDelimiter);
+    }
```
Could the user not just pass `Model.schema()` to the relation setup call instead?

mliszewski replied:
They could but then that would require the user to setup a separate association for every schema that they need to access this association in with different names. In an ideal world I can setup one set of models and associations and use them across all schemas without having to define the models and associations again with different names for each schema. That being said it would work as you described as well if that use case make more sense for the user than you would just not pass the schema option and it would use the schema associated with the model when the association was setup.

mickhansen replied:
Makes sense for dynamic schemas.

mickhansen commented on lib/instance.js:
```diff
@@ -640,16 +641,24 @@ Instance.prototype.save = function(options) {
+      if (self.$options.$schema !== self.Model.$schema ||
+        self.$options.$schemaDelimiter !== self.Model.$schemaDelimiter) {
+        tableName = self.Model.schema(self.$options.$schema, self.$options.$schemaDelimiter).getTableName(options);
+      } else {
```
Why is this code needed exactly?
Seems odd there should be a case where there's mismatch in the instance model?

mliszewski replied:
This is not odd at all if I am trying to sue the same model across schemas. The model is attached to the Instance prototype when init is called on the model: https://github.com/sequelize/sequelize/blob/master/lib/model.js#L724-L725. If I then call .schema on this model it will not change $schema on the model attached to the prototype.

mickhansen replied:
Hmm alright, i was thinking of doing a full clone on schema so the instance would be cloned aswell.

mliszewski replied:
Yeah I think the issue with trying to do this (at least from what I could tell) is that since Model/$Model are set on the prototype of the instance then changing the schema on one instance will change it on all which causes this test case to fail: https://github.com/sequelize/sequelize/pull/4599/files#diff-7de0096efac48f1ebb1ecedda103ca05R362.

mickhansen replied:
You'd have to first clone the model, then clone Model.Instance and then set schema properties.

mliszewski replied:
But if you clone Model.Instance using Object.create (which is ideal from a resource perspective) it will not matter b/c the Model property of instance is set on the prototype. I guess you could detach the Model from the prototype but this will cause all kinds of memory bloat. I guess you could also create attributes to specifically hold the schema properties on the Instance object and set them but that seems like a worse solution that my current solution that will read the $schema attribute from $options which sets them based on the model the actual instance was built from (https://github.com/sequelize/sequelize/pull/4599/files#diff-a140b60b6a99a8c69f4b5d0d1f1f2bfaL1706) which seems like the expected behavior.

mickhansen replied:
Not sure i understand, couldn't we just:

```js
var clone = Object.create(Model);
clone.instance = function() {
   Model.Instance.apply(this, arguments);
};
clone.Instance.prototype = Object.clone(Model.Instance.prototype);
clone.Instance.prototype.$Model = clone;
```

`Model.schema` should perhaps then have a cache, LRU perhaps.
I would greatly prefer it to having 3-4 places with special handling code.

/cc @janmeier

janmeier replied:
Yeah, making a clone of the instance makes more sense to me a well

mliszewski commented:
@mickhansen Let me know if my explanations make sense and if you would like to see anything else changed or explained. Thanks.

mliszewski commented:
@mickhansen @janmeier I changed the code to clone instance as you recommended. Tests passed for me locally but for some reason they haven't picked up in CI yet. Can you run tests and give me any additional feedback when you get a chance? Thanks

mickhansen commented:
Hmm looks like it didn't pick up the rebase.

mickhansen commented:
This looks really great now @mliszewski - Not sure why it's not running the tests though, any ideas @janmeier?

janmeier commented:
@mickhansen opening and closing the PR fixed it ;).

This LGTM unless you have further comments?
